ports: Support UART.irq for the RP2, ESP32, MIMXRT, NRF, Renesas-RA and SAMD ports.

robert-hh opened the pull request:
Adds UART IRQ functionality for 4 additional ports: RP2, ESP32, MIMXRT and SAMD. Available triggers:

IRQ_RX: called every time when a byte is received. Not useful for high data rates (like > 9600 baud).
IRQ_RXIDLE: called when the receive line gets silent after a burst of data. Very useful!
IRQ_TXIDLE: called when a block of data has been sent.

Not all ports provide all triggers:

- rp2: IRQ_RXIDLE, IRQ_TXIDLE and IRQ_BREAK. IRQ_RXIDLE works as it should. IRQ_TXIDLE is only called when the data length is mor than 5 bytes, and it triggers when still 4-5 bytes are to be sent.
- MIMXRT: IRQ_RXIDLE and IRQ_TXIDLE. Work as they should with a fix to the NXP UART driver. which is included in this PR.
- SAMD: IRQ_RX and IRQ_TXIDLE. The SAMD port's hardware does not support IRQ_RXIDLE. Based on the IRQ_RX trigger a simple class can be made which implements a IRQ_RXIDLE handler call. IRQ_TXIDLE triggers while the last byte is sent.
- ESP32: IRQ_RX and IRQ_BREAK: The ESP32 does not support IRQ_RXIDLE. The option `hard=True` is not supported.
- NRF: IRQ_RX and IRQ_TXIDLE. Enabled for boards with nrf52480. I could not test other boards.

Tested with RP2040, ESP32. ESP32-S2, ESP32-S3, ESP32-C3, SAMD21, SAMD51, NRF52840, and all supported MIMXRT MCUs except i.MX RT 1064 (I do not have access to a board).

Test observations for other boards:

- STM32: IRQ_RXIDLE. For F4xx MCU (e.g. Pyboard), the RXIDLE interrupt triggers twice with every message longer than 1 byte starting with the second message after UART init. The first trigger happens at the first byte of the message, the second after the message. 1 Byte messages get only one trigger, but with different delays after the message. Tests with F7xx and H7xx boards showed proper behavior.
- Renesas-RA: Trigger IRQ_RX.
- CC2300: Trigger RX_ANY. IRQ_RX was added as alias in this PR.

robert-hh committed:
mimxrt/machine_uart: Implement a Python UART IRQ handler.

Supported triggers are: IRQ_RXIDLE and IRQ_TXIDLE.

When IRQ_RXIDLE is set, the handler will be called 3 character times after
the data in burst stopped.

When IRQ_TXIDLE is set, the handler will be called immediately after the
data has been sent.

This commit requires a change to fsl_lpuart.c, because the existing code
does not support under-run appropriately.

The irq.flags() value is cleared only at an expected event.  Do not change
it otherwise.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
rp2/machine_uart: Implement a Python UART IRQ handler.

Supported trigger names: IRQ_RXIDLE, IRQ_TXIDLE, IRQ_BREAK

- IRQ_RXIDLE: The handler for IRQ_RXIDLE is called reliably 31 UART bit
  times after the last incoming data.

- IRQ_TXIDLE: This IRQ is triggered after at least >5 characters are sent
  at once.  It is triggered when the TX FIFO falls below 4 elements.  At
  that time, up to 5 bytes may still be in the FIFO and output shift
  register.

- IRQ_BREAK: The IRQ triggers if a BREAK state is detected at RX.
  Properties & side effects:
  - After a BREAK, a valid character must be received before another break
    can be detected.
  - Each break puts a 0xff character into the input buffer.

The irq.flags() value is cleared only with a new wanted event.  Do not
change the flags otherwise.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/machine_uart: Implement a Python UART IRQ handler.

Supported for all SAMD51 devices and SAMD21 with external flash.  For
interrupt events, IRQ_RX and IRQ_TXIDLE are provided.

IRQ_RX is called for every received byte.  This may not be useful for high
data rates, but can be used to build a wrapper class providing an
IRQ_RXIDLE event or to signal just the first byte of a message.

IRQ_TXIDLE is called only when messages are longer than 5 bytes and
triggers when still 5 bytes are due to be sent.

The SAMD hardware does not support implementing IRQ_RXIDLE.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge committed:
docs/library/machine.UART: Fix UART.irq docs to match current code.

These docs now match the code in `extmod/machine_uart.c`.  IRQ trigger
support still need to be updated for each port (to be done in a follow-up
commit).

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Thanks for this, it's a great addition!  Definitely needed.

By chance I was working on the docs side of this today: #14040.

robert-hh committed:
renesas-ra/machine_uart: Add the UART.IRQ_RX class constant.

The renesas-ra port supports calling a handler to be called on every byte
received by UART.  For consistency with other ports, the symbol IRQ_RX
is added as the trigger name.

Side change: Add the received UART data to the REPL input buffer only if it
is the REPL UART.  Otherwise, every UART would act as REPL input.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
cc3200/mods/pybuart: Add the UART.IRQ_RX class constant.

As alternative to RX_ANY to match the names used by the other ports.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
esp32/machine_uart: Implement Python UART IRQ with IRQ_RX and IRQ_BREAK.

Supported trigger events: IRQ_RX and IRQ_BREAK.  Hard IRQ is not supported.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
docs/library/machine.UART: Extend the documentation for UART.irq.

For more ports and trigger options, based on the current state of the code.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
nrf/modules/machine/uart: Implement Python UART IRQ for nrf52840 boards.

Supported triggers: UART.IRQ_RX and UART.IRQ_TXIDLE.  It will probably work
on other boards as well, but so far untested.

The irq.flags() value is changed only when requested by a triggered event.
Do not change it otherwise.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
A few discussion topics:
- Some alternative names for the triggers:
  IRQ_RXANY instead of IRQ_RX
  IRQ_TXDONE instead of IRQ_TXIDLE

- All but the ESP32 port have the same section of code now, which implement the UART.irq  method. I wonder if that could not be moved for example into extmod/machine_uart.c.

robert-hh committed:
stm32/machine_uart: Add the UART.IRQ_RX event for UART.irq().

Just adding the event symbol.  No code change required, and no impact on
code execution time when the event is not selected.

Tested with STM32F4xx, STM32F7xx and STM32H7xx.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented:
Thanks again for working on this.  After thinking more about it here are my general comments.

I don't think it makes sense to have both `IRQ_RX` and `IRQ_RXIDLE`, especially because it seems that the ports either implement one or the other, but never both.

We need to make it as easy as possible to write portable code that runs across all the ports, and having different IRQs, or the same IRQ but with slightly different semantics, makes that hard.

So I suggest to not have `IRQ_RX` at all, and make `IRQ_RXIDLE` mandatory.  Users should be able to rely on thisIRQ to have the following semantics:

    IRQ_RXIDLE: every character received by the UART will have this interrupt
    triggered at some point in time after that character is received

The idea is that if you only ever do `uart.read()` within the `IRQ_RXIDLE` callback, then you should always be able to (eventually) read all available characters.  And ideally in a timely manner.

Even if the hardware/SDK doesn't support this concept we should still be able to synthesize it.  Eg if the hardware supports `IRQ_RX` then in combination with a timer we can implement `IRQ_RXIDLE`.

Having `IRQ_BREAK` (when the hardware supports it) makes sense, the semantics for that seem obvious.

For `IRQ_TXIDLE`/`IRQ_TXDONE`: as with `IRQ_RXIDLE` the semantics of this need to be well specified, and then all ports should implement it.

robert-hh commented:
So you suggest that IRQ_RX is renamed to IRQ_RXIDLE for those ports that do not support a true IRQ_RXIDLE, and for those that do so IRQ_RX is dropped?

B.t.w: I made two Python classes as example that implement IRQ_RXIDLE based on an existing IRQ_RX. The simpler one is below with uart.irq() being a method. The other implements uart.irq() as a class. I considered implementing that in the machine_uart module itself, but dealing with the timer is easier in Python.
```
#
# Simple wrapper class for UART, adding the trigger IRQ_RXIDLE for port
# which just support IRQ_RX
# Missing: Handling of the irq.xxx methods in this class.
#
import machine

class UART(machine.UART):
    def __init__(self, id, *args, **kwargs):
        self.irq_timeout_count = 0
        self.irq_rx_count = 0
        self.timeout_idle = 5
        self.irq_timer = None
        self.irq_handler = None
        self.IRQ_RXIDLE = 4096

    def irq(self, handler=None, trigger=None, hard=False, period=5):
        if handler is not None:
            if trigger == self.IRQ_RXIDLE:
                self.irq_handler = handler
                self.irq_timer = machine.Timer(callback=self.irq_timer_callback,
                    mode=machine.Timer.PERIODIC, period=period)
                return super().irq(handler=self.irq_rx_callback, trigger=self.IRQ_RX, hard=True)
            else:
                if self.irq_timer is not None:
                    self.irq_timer.deinit()
                self.irq_handler = None
                return super().irq(handler=handler, trigger=trigger, hard=hard)
        else:
            if self.irq_timer is not None:
                self.irq_timer.deinit()
            self.irq_handler = None
            return super().irq(handler=None, trigger=0, hard=False)

    def irq_rx_callback(self, uart):
        self.irq_rx_count += 1
        self.irq_timeout_count = 0

    def irq_timer_callback(self, timer):
        if self.irq_rx_count > 0:
            self.irq_timeout_count += 1
            if self.irq_timeout_count == 2:
                if self.irq_handler is not None:
                    self.irq_handler(self)
                self.irq_rx_count = 0
                self.irq_timeout_count = 0
```

robert-hh commented:
@dpgeorge Thinking about further, I prefer to support the behaviour with proper names. Looking at the list of ports, three of them support true IRQ_RXIDLE, and eight support IRQ_RX. Both the rp2040 and the MIMXRT port can support IRQ_RX mode. The rp2040 commit would get much simpler and more close to the existing state. Without support for IRQ_RXIDLE, the MIMXRT port would not need a change to the NXP UART driver. 
So I suggest to have IRQ_RX as the common method and have support for IRQ_RXIDLE as a derived class method.

dpgeorge commented:
> So I suggest to have IRQ_RX as the common method and have support for IRQ_RXIDLE as a derived class method.

I understand it's simpler to implement what the hardware provides.  But:
- IRQ_RX is not that useful for large data rates (as you already mentioned).  So it will be impossible to implement IRQ_RXIDLE behaviour in Python using IRQ_RX (because IRQ_RX will be called too fast for Python to keep up with it).  Even if the implementation does work, it won't be efficient.
- We need to provide something portable that works the same across all ports.  Eg so the `network.PPP` driver can work and use the UART IRQ.

robert-hh commented:
For the existing PR is see then two options:
a) Drop the changes for IRQ_RX and keep only the changes for RP2040 and MIMXRT for IRQ_RXIDLE. More boards may be added later it a IRQ_RXIDLE is implemented based on an internal IRQ_RX and a timer. Maybe the easiest way would be to compare the ms time stamp at an char RX with that of the previous char RX.
b) Rename IRQ_RX to IRQ_RXIDLE and have a note in the documentation about the difference to real IRQ_RXIDLE.

Obviously the behavior of RXIDLE can as well be implemented in Python with a timer interrupt, which checks if uart.any() values stall at a value > 0.

projectgus reviewed (commented):
Left a couple of small comments in the docs, and agree about thinking through the most useful RX interrupt behaviour, but overall this looks very useful. Thanks for adding it on so many ports, @robert-hh!

dpgeorge commented:
After discussion with @jimmo and @projectgus we came to the conclusion that it makes sense to expose multiple UART IRQs, and then the user can decide which is best for their application.  But we do need to have consistence across ports for the behaviour of the interrupts.

- IRQ_RX: all ports will be able to implement this in hardware (I assume?) and it's a pretty useful interrupt to have access to, although only for slow baud rates.
- IRQ_RXFIFO: for when the RX input buffer is half full.  This can be easily synthesized (in C) as long as IRQ_RX is available.  Note that we don't need to implement this one yet but there is scope to add it in the future.
- IRQ_RXIDLE: when the RX line remains high for whole character period after receiving at least one character.  This can be synthesized (in C) for ports that don't have hardware support for it.
- IRQ_BREAK: useful and obvious semantics.
- IRQ_TXIDLE: not so sure about this one, it's a bit harder to define and we don't yet have consistency across ports for TX, whether it's blocking or runs in the background.

So in the end, based on the above points, I think the PR as it stands is in pretty good shape!  The main thing is to have good documentation that specifies the behaviour of each IRQ well, and makes sure the implementations for each port match that behaviour so it can be relied on.

robert-hh commented on docs/library/machine.UART.rst:
```diff
@@ -203,11 +178,85 @@ Methods
+       ``UART`` instance.
+
+     - *trigger* configures the event which can generate an interrupt.
+       Possible values are:
```
Thanks. Updated accordingly.

robert-hh commented on docs/library/machine.UART.rst:
```diff
@@ -203,11 +178,85 @@ Methods
+       - ``UART.IRQ_RX`` interrupt after each received character.
+       - ``UART.TX_IDLE`` interrupt after or while the last character(s) of
+         a message are or have been sent.
+       - ``UART.BREAK`` interrupt when a break state is detected at RX
```
Thanks. Updated accordingly. In the code, these names were used.

robert-hh commented on docs/library/machine.UART.rst:
```diff
@@ -203,11 +178,85 @@ Methods
+      :align: center 
+
+      ============== ======= === ====== =====
+      Port / Trigger RX_IDLE RX  TXIDLE BREAK
```
Thanks. Updated accordingly

robert-hh commented:
Thank you for the confirmation. Some notes:

> IRQ_RX: all ports will be able to implement this in hardware

Actually, at the RP2040 and MIMXRT port one has to chose between IRQ_RX and IRQ_RXIDLE. For the RP2040, it's the way interrupts are created and the FIFO is working. Unfortunately it is not possible to tell how many bytes are in the FIFO. And IRQ_TXIDLE happens only if at least one byte is in the FIFO. For that reason, the FIFO full IRQ is set to 75% (24 Byte), and in an interrupt not more than 23 bytes are removed from the FIFO. read() takes all data from the FIFO. but with an IRQ_RX at 24Byte this is not so useful. The MIMXRT UART driver has no IRQ_RX event, since that is all handled in the driver itself. But I can check again if some of the other Event match.

> IRQ_RXFIFO: for when the RX input buffer is half full.

That looks doable to some extend for most ports, if the RX ring buffer is taken as reference.

> IRQ_RXIDLE: when the RX line remains high for whole character period after receiving at least one character. This can be synthesized (in C) for ports that don't have hardware support for it.

Principally yes. The challenge will be to use the right timer mechanism with little overhead. The time the RX line has to stay high is 30 bit times for RP2040, which is hardwired, and for MIMXRT I have configured it to 4 byte times starting from the most recent start bit, resulting in about 30 bit times after the last byte. If I recall right, the STM32 timing is similar.
The Python code above works well for slow communication like that from a GPS receiver or other sensors. So it has it's use case.

> IRQ_BREAK: useful and obvious semantics.

I looked at the datasheets for all eight ports. For only two ports I found a hardware interrupt or signalling mechanism for break. I'm still wondering why this is not available at the MIMXRT or STM32 port.

> IRQ_TXIDLE: not so sure about this one, it's a bit harder to define and we don't yet have consistency across ports for TX, whether it's blocking or runs in the background.

Yes. This one should happen when all data of a "message" is sent. But yes, if the driver buffers internally without being able to tell, that could be tricky. And we have already uart.flush() and uart.txdone(), which could be used for that purpose, with the same limitation.

I updated the documentation according to Angus' comments.

robert-hh committed:
samd/machine_uart: Implement UART.IRQ_RXIDLE based on the softtimer.

With the softtimer the minimal delay between the end of a message and the
trigger is 2 ms.  For baud rates <= 9600 baud it's three character times.
Tested with baud rates up tp 115200 baud.  The timer used for RXIDLE is
running only during UART receive, saving execution cycles when the timer is
not needed.

The irq.flags() value is changed only with an expected event.  Do not
change it otherwise.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
The PR is now updated with a  sample implementation of IRQ_RXIDLE for the SAMD port using softtimer. The same approach can be used for the Renesas port. 
For that purpose, the softtimer data structure is extended by an additional pointer, which can port at context data for the timer interrupt. In this case it's the respective UART data structure. That allows to have a single timer callback and not a separate callback for each UART. At the moment that does not interfere with any other use of softtimer. To make is consistent, the function softtimer_static_init() may get another argument, the context pointer. That requires to change all call of that function. Not a big thing, but I wanted to wait until the approach of a context pointer is accepted.

robert-hh committed:
renesas-ra/machine_uart: Implement UART.IRQ_RXIDLE based on softtimer.

Allowing to define the trigger UART.IRQ_RXIDLE as well as UART.IRQ_RX.  The
delay for the IRQ_RXIDLE interrupt is about 3 character times or 1-2 ms,
whichever is larger.

The irq.flags() value is changed only with an expected event.  Do not
change it otherwise.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
esp32/machine_uart: Implement UART.RX_IDLE based on machine.Timer.

The UART.IRQ_IDLE callback is called about two character times after the
last byte, or 1 ms, whichever is larger.  For the irq, timer 0 is used.

machine_timer.c had to be reworked to make it's mechanisms available for
machine_uart.c.

The irq.flags() value is change only at a requested event.  Otherwise keep
the state.

Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented on shared/runtime/softtimer.h:
```diff
@@ -44,6 +44,7 @@ typedef struct _soft_timer_entry_t {
         void (*c_callback)(struct _soft_timer_entry_t *);
         mp_obj_t py_callback;
     };
+    void *context;
```
To avoid adding this to all users of this struct it might be possible to instead create a derived struct like this:
```c
typedef struct _soft_timer_entry_extended_t {
    soft_timer_entry_t base;
    void *context;
} soft_timer_entry_extended_t;
```
Then you can pass `&entry->base` to the existing soft-timer functions, but when you get a callback you can cast the `soft_timer_entry_t*` back to a `soft_timer_entry_extended_t*` and access the context.

robert-hh replied:
Thanks for the hint. The code is changed now and tested with the SAMD and Renesas port, which are affected by the change.
I could not test with the Pico2, since I have no board yet. But when adding the RP2 commit to the RP2350 PR, it still builds fine.

dpgeorge commented on ports/nrf/boards/ARDUINO_NANO_33_BLE_SENSE/mpconfigboard.h:
```diff
@@ -17,6 +17,7 @@
 #define MICROPY_PY_MACHINE_I2C          (1)
 #define MICROPY_PY_MACHINE_ADC          (1)
 #define MICROPY_PY_MACHINE_TEMP         (1)
+#define MICROPY_PY_MACHINE_UART_IRQ     (1)
```
Is there a way to automatically set this option in `mpconfigport.h`, so it doesn't need to be done per board?  For example, all nRF52 boards get it, or use `EXTRA_FEAT`.

robert-hh replied:
It can be based on the #define of `NRF52`. But there is also a define for `MICROPY_PY_MACHINE_UART` in every board file. That could be moved to mpconfigport.h as well, especially since the uart includefile is set in mpconfigport.h.

dpgeorge replied:
> But there is also a define for `MICROPY_PY_MACHINE_UART` in every board file.

Right... ideally that would be moved to `mpconfigport.h`, but maybe that can be left for another PR, to clean these up.

Do as you see fit.

robert-hh replied:
I'll leave that for a separate clean-up. There is more to re-arrange.

dpgeorge commented on docs/library/machine.UART.rst:
```diff
@@ -203,11 +178,85 @@ Methods
+       idle.
+
+
+   Availability: renesas-ra, stm32, esp32, rp2040, mimxrt, samd, nrf, cc3200.
```
Please sort these in alphabetical order.

robert-hh replied:
Done

dpgeorge committed:
tests/extmod: Add test for machine.UART.IRQ_TXIDLE.

The test checks whether the message created by the IRQ handler appears
about at the end of the data sent by UART.

Supported MCUs resp. boards:
- RP2040
- Teensy 4.x
- Adafruit ItsyBitsy M0
- Adafruit ItsyBitsy M4
- NRF52 (Arduino Nano Connect 33 BLE)

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod_hardware: Add tests for machine.UART.IRQ_RX/RXIDLE/BREAK.

These all require hardware connections, so live in a different directory.

Except for the IRQ_BREAK test of ESP32 devices a single UART with loopback
is sufficient.

General:
    SAMD21: Due to the limited flash size only SAMD21 devices with external
    flash support uart.irq().

IRQ_BREAK:
    ESP32 needs different UART devices for creating and sensing a break.
    Lacking a second UART the test is skipped for ESP32S2 and ESP32C3.  RP2
    does not pass the test reliable at 115200 baud, reason to be found.
    Thus the upper limit is set to 57600 Baud.

    Coverage:
        esp32  pass when different UART devices are used.
        rp2    pass up to 57600 baud

IRQ_RX:
    SAMD21: Being a slow device it needs data to be sent byte-by-byte at
    9600 baud, since the IRQ callback is scheduled delayed and then the
    flags do not match any more.  The data matches since it is queued in
    the FIFO resp. ringbuffer.

    CC3200: The test cannot be performed since no calls are accepted in the
    IRQ handler like u.read(). Skipped.

    Coverage:
        cc3200 fail due to major differences in the implementation.
        esp32  pass
        nrf    pass
        renesas-ra pass
        samd   pass see the notes.
        stm32  pass

IRQ_RXIDLE:
    STM32: With PyBoard the IRQ is called several times, but only once with
    the flag IRQ_RXIDLE set.

    Coverage:
        esp32    pass
        mimxrt   pass
        renesas-ra pass
        rp2      pass
        samd     pass for both SAMD21 and SAMD51
        stm32    fail. see notes.

Signed-off-by: Damien George <damien@micropython.org>
Signed-off-by: robert-hh <robert@hammelrath.com>

dpgeorge commented:
@robert-hh I have added tests for all the IRQs added in this PR.  I pushed those tests to your branch.  Feel free to modify the tests, eg add more configurations for other ports.

Each test must have configuration for a given port, eg the UART id and UART pins.  Some tests can run without extra hardware connections, but others require connecting TX and RX.  The tests are documented at their top to say what they need.

I ran the tests on a few different ports (stm32, rp2, esp32, samd).  Not all of them pass.  It would be great if all tests could pass on all ports (or skip if the port doesn't support that IRQ), but I know that could be difficult to achieve.  If you have some time to look into making them pass that would be fantastic!

--------

For example, running the tests on various hardware:

#### ESP32

```
$ ./run-tests.py --target minimal --device /dev/ttyUSB0 extmod/machine_uart*.py extmod_hardware/machine_uart*.py
skip  extmod/machine_uart_irq_txidle.py
skip  extmod/machine_uart_tx.py
FAIL  extmod_hardware/machine_uart_irq_break.py
pass  extmod_hardware/machine_uart_irq_rxidle.py
pass  extmod_hardware/machine_uart_irq_rx.py
3 tests performed (36 individual testcases)
2 tests passed
2 tests skipped: machine_uart_irq_txidle machine_uart_tx
1 tests failed: machine_uart_irq_break
```

#### SAMD21

```
$ ./run-tests.py --target minimal extmod/machine_uart*.py extmod_hardware/machine_uart*.py             
FAIL  extmod/machine_uart_irq_txidle.py
skip  extmod/machine_uart_tx.py
skip  extmod_hardware/machine_uart_irq_break.py
pass  extmod_hardware/machine_uart_irq_rxidle.py
FAIL  extmod_hardware/machine_uart_irq_rx.py
3 tests performed (30 individual testcases)
1 tests passed
2 tests skipped: machine_uart_irq_break machine_uart_tx
2 tests failed: machine_uart_irq_rx machine_uart_irq_txidle
```

#### rp2 (Pico)

```
$ ./run-tests.py --target minimal extmod/machine_uart_irq*.py extmod_hardware/machine_uart_irq*.py 
FAIL  extmod/machine_uart_irq_txidle.py
pass  extmod_hardware/machine_uart_irq_break.py
pass  extmod_hardware/machine_uart_irq_rxidle.py
skip  extmod_hardware/machine_uart_irq_rx.py
3 tests performed (33 individual testcases)
2 tests passed
1 tests skipped: machine_uart_irq_rx
1 tests failed: machine_uart_irq_txidle
```

#### stm32 (PYBv1.0)

```
$ ./run-tests.py --target minimal extmod/machine_uart_irq*.py extmod_hardware/machine_uart_irq*.py
skip  extmod/machine_uart_irq_txidle.py
skip  extmod_hardware/machine_uart_irq_break.py
FAIL  extmod_hardware/machine_uart_irq_rxidle.py
pass  extmod_hardware/machine_uart_irq_rx.py
2 tests performed (21 individual testcases)
1 tests passed
2 tests skipped: machine_uart_irq_break machine_uart_irq_txidle
1 tests failed: machine_uart_irq_rxidle
```

robert-hh commented:
I'll see what can be done for testing. In my tests I used two modules. Using a single device with a kind of loopback was sometimes not possible..

robert-hh commented:
The test scripts are now reworked to some extend. Tests pass for almost all supported boards. Some boards cannot pass the test even if the feature works:
- ESP32S2 and ESP32C3 have no UART2 and cannot run the IRQ_BREAK test. IRQ_BREAK cannot be performed in a UART loopback mode. IRQ_BREAK works an can be triggered with a GPIO pulse, but then there is no serial data as expected by the test.
- STM32 (PyBoard) fails at the IRQ_RXIDLE test even if the mechanism works. It just triggers more than one IRQ, and then the state of the flag and the expected data mismatch.
- CC3200 is too slow for the IRQ_RX test, and it locks up in the callback when attempting to allocate RAM. Given that the CC3200 is legacy, the test support is not included.

dpgeorge commented:
> The test scripts are now reworked to some extend. Tests pass for almost all supported boards.

Excellent, thanks for working on that!  I will test them myself.

robert-hh commented:
Re-testing this morning when running all tests in a row with the loopback jumper attached I see a few hiccups:
- the NRF port locks up when trying to run the second test. It seems to require a hard reset between tests.
- SAMD fails the IRQ_TXIDLE test with the loopback jumper attached. In that case, the IRQ callback is called twice instead of once. There is a similar multiple call situation with STM32 and IRQ_RXIDLE.

robert-hh commented:
I updated the SAMD port such that the double callbacks are gone.

Gadgetoid commented:
Looking at RP2 (regular Pico), trying to send 512 bytes I get multiple IRQs firing but only 288 bytes of data in the first, and nothing in the subsequent IRQ(s). Is this normal, related to my issues with LTE/PPP?

```python
from machine import UART
import random
import time, sys


def irq(u):
    if u.irq().flags() & u.IRQ_RXIDLE:
        data = u.read()
        if data:
            print("IRQ_RXIDLE:", bool(u.irq().flags() & u.IRQ_RXIDLE), "data:", data, len(data))
        else:
            print("IRQ_RXIDLE: NO DATA!?")


BUF_LEN = 512
text = "".join(chr(random.randint(ord('A'), ord('z'))) for _ in range(BUF_LEN))
 
uart = UART(0, 115200, tx=0, rx=1, timeout=1)
uart.irq(irq, uart.IRQ_RXIDLE)

while True:
    print(f"\nWriting {len(text)} bytes...")
    uart.write(text)
    time.sleep_ms(5000)
```

eg:

```
Writing 512 bytes...
IRQ_RXIDLE: True data: b'b]Ubf_s`zOGscbZvdifV^FFsZyxK^gXh_yOjxHTYwOJOCiPyYfV^RYKx`ZPN[eBbDACGjumymFqvLi^X[swHFBGucRZjUkdNuiDixdpidqTDyD_W^rPngQIkkMFmJwycUdgpzgMwWW]LSMYtzgGpjCznIhP[Kpg_]Qa]TLxwmcRDDRjDJqkCVhUZ]AijGRuWwY_iUHqOimvXdbqEfgKHwvDlWpvsWFuhBwZsTV]vGycxutlZXZI\\HQe\\UmAbFeefjZZPmONI[ai]ImmDXWyCpxXedF]rZkHG' 288
IRQ_RXIDLE: NO DATA!?
IRQ_RXIDLE: NO DATA!?

Writing 512 bytes...
IRQ_RXIDLE: True data: b'b]Ubf_s`zOGscbZvdifV^FFsZyxK^gXh_yOjxHTYwOJOCiPyYfV^RYKx`ZPN[eBbDACGjumymFqvLi^X[swHFBGucRZjUkdNuiDixdpidqTDyD_W^rPngQIkkMFmJwycUdgpzgMwWW]LSMYtzgGpjCznIhP[Kpg_]Qa]TLxwmcRDDRjDJqkCVhUZ]AijGRuWwY_iUHqOimvXdbqEfgKHwvDlWpvsWFuhBwZsTV]vGycxutlZXZI\\HQe\\UmAbFeefjZZPmONI[ai]ImmDXWyCpxXedF]rZkHG' 288
IRQ_RXIDLE: NO DATA!?
```

robert-hh commented:
I'll check. Did you have a loopback connection at UART 0?

robert-hh commented:
You have to set rxbuf to a larger value. The default is 256. It should be 512 at least.

Gadgetoid commented:
> The default is 256. It should be 512 at least.

Hmm I suppose I should probe deeper into what my LTE module tries to send, since this implies it could be overwhelming the rx buffer and might explain why I must use polling.

That said, a generously large (4k) rxbuf does not appear to help. Which only really suggests this is unrelated.

dpgeorge commented on tests/extmod_hardware/machine_uart_irq_break.py:
```diff
@@ -0,0 +1,67 @@
+
+# Configure pins based on the target.
+if "esp32" in sys.platform and not (
+    "ESP32S2" in sys.implementation._machine or "ESP32C3" in sys.implementation._machine
```
I suggest to SKIP s2/s3 like this:
```py
if "esp32" in sys.platform:
    if "ESP32S2" in sys.implementation._machine or "ESP32C3" in sys.implementation._machine:
        print("SKIP")
        raise SystemExit
    ... set up pins for esp32
elif "rp2" in sys.platform:
    ... set up pins for rp2
else:
    print("Please add support for this test on this platform.")
    raise SystemExit
```
That way any new ports in the future that support IRQ_BREAK have an appropriate warning about adding support to this test.

dpgeorge commented on tests/extmod_hardware/machine_uart_irq_break.py:
```diff
@@ -0,0 +1,67 @@
+if "esp32" in sys.platform and not (
+    "ESP32S2" in sys.implementation._machine or "ESP32C3" in sys.implementation._machine
+):
+    # The ESP32 needs separate UART instances for the test
```
It's a shame that esp32 needs two instances... I see why, it's because sendbreak() changes the TX and RX baudrate.

I think we should change sendbreak() so it pulls the TX GPIO low for the correct period of time.  Otherwise in real applications one could lose RX chars while doing a sendbreak.  (But this is for a separate PR in the future.)

robert-hh replied:
OK

dpgeorge commented on tests/extmod_hardware/machine_uart_irq_break.py:
```diff
@@ -0,0 +1,67 @@
+
+# Test that the IRQ is called for each break received.
+for bits_per_s in (2400, 9600, 57600):
+    if recv_tx_pin is None:
```
This block of if-code is not needed.

robert-hh replied:
I kept it for a possible future board that does not allow specifying the rx/tx pins. I still wonder why the MIMXRT lib does not provide the `break` event.

dpgeorge replied:
I would rather go the other way and make sure all ports allow specifying the tx/rx pins (not required but allowed).  That makes ports more consistent.

For a port that can't change the tx/rx pin then all it needs to do is verify that the passed-in pin matches the pre-configured pin.

Anyway, all that is for another day.  For now I don't think we need this bit of code in the test.

robert-hh replied:
Removed

dpgeorge commented on tests/extmod_hardware/machine_uart_irq_break.py:
```diff
@@ -0,0 +1,67 @@
+        time.sleep_ms(10)
+        send_uart.sendbreak()
+        time.sleep_ms(10)
+        recv_uart.read()
```
It seems this read is only needed for esp32, so I suggest doing this:
```py
if "esp32" in sys.platform:
    # On esp32 a read is needed to read in the break byte.
    recv_uart.read()
```

robert-hh replied:
If there is no data, the read returns `None`. So it does not hurt.

robert-hh replied:
Done

dpgeorge commented on tests/extmod_hardware/machine_uart_irq_rx.py:
```diff
@@ -0,0 +1,78 @@
+    uart.irq(irq, uart.IRQ_RX)
+
+    print("write", bits_per_s)
+    if byte_by_byte:
```
I suggest to just always write byte-by-byte.  That doesn't really change the test for those ports that are fast enough.

robert-hh replied:
But then the test would not reflect the real operation. Better then to skip testing slow devices like SAMD21.

robert-hh replied:
Then the .exp file does not match.

dpgeorge replied:
That's true...

OK, then let's leave it as you have it.

dpgeorge commented on tests/extmod_hardware/machine_uart_irq_rx.py:
```diff
@@ -0,0 +1,78 @@
+    tx_pin = None  # P602 @ RA6M2
+    rx_pin = None  # P601 @ RA6M2
+else:
+    print("SKIP")
```
I suggest keeping this line as `print("Please add support for this test on this platform.")`, because the test should not really be skipped on boards that do support this feature.

robert-hh replied:
Done. But we know it will fail for the CC3200, since it supports IRQ_RX but cannot pass the test.

dpgeorge replied:
OK, for cc3200 it can have an explicit section here that skips that port.  And a comment.  Eg:
```py
elif sys.platform == "cc3200":
    # CC3200 doesn't work because it's too slow and has an allocation error in the handler.
    print("SKIP")
    raise SystemExit
```

robert-hh replied:
Done.

dpgeorge commented on tests/extmod_hardware/machine_uart_irq_rxidle.py:
```diff
@@ -0,0 +1,71 @@
+    tx_pin = "D1"
+    rx_pin = "D0"
+else:
+    print("SKIP")
```
I suggest keeping `print("Please add support for this test on this platform.")`.

robert-hh replied:
Done

dpgeorge commented on tests/extmod_hardware/machine_uart_irq_rxidle.py:
```diff
@@ -0,0 +1,71 @@
+
+
+def irq(u):
+    if u.irq().flags() & u.IRQ_RXIDLE:
```
Is stm32 the only port that needs this check?  If so then I suggest to remove the check and let the stm32 port fail.  That will serve as a reminder to fix stm32 (in a separate PR).

robert-hh replied:
Removed.

dpgeorge commented on tests/extmod/machine_uart_irq_txidle.py:
```diff
@@ -0,0 +1,69 @@
+    tx_pin = None
+    min_window = 0
+else:
+    print("SKIP")
```
I suggest keeping `print("Please add support for this test on this platform.")`.

robert-hh replied:
Done

dpgeorge commented on tests/extmod/machine_uart_irq_txidle.py:
```diff
@@ -0,0 +1,69 @@
+
+
+def irq(u):
+    print("IRQ_TXIDLE:", bool(u.irq().flags() & u.IRQ_TXIDLE))
```
Can this be a `u.irq().flags() == u.IRQ_TXIDLE`?  Which ports require the `&`?

robert-hh replied:
I just made that identical to the other test scripts, which use the `bool(a & b)` term.

dpgeorge replied:
It's stricter to test with `==`, which I would prefer to do if possible.

The tests that require `bool(a & b)` probably need it for a (good?) reason.

robert-hh replied:
Changed back for IRQ_TXIDLE.

dpgeorge commented on tests/extmod/machine_uart_irq_txidle.py:
```diff
@@ -0,0 +1,69 @@
+
+    bits_per_char = 10  # 1(startbit) + 8(bits) + 1(stopbit) + 0(parity)
+    start_time_us = (len(text) - 10) * bits_per_char * 1_000_000 // bits_per_s
+    window_ms = max(min_window, 20 * bits_per_char * 1_000 // bits_per_s + 1)
```
It's probably worth commenting on the algorithm used in these two calculations.  Why `-10` for `start_time_us`, and why `20 *` for `window_ms`, etc?

robert-hh replied:
Added. The value of 20 could probably be smaller.  SAMD devices use busy waiting for sleep_ms(n)
for n <= 10 and for sleep_us(), preventing the IRQ to get scheduled. For these I added the minimum wait time. But I prepared a change for SAMD to make that obsolete. Will be in a separate PR.

dpgeorge commented on docs/library/machine.UART.rst:
```diff
@@ -203,11 +178,85 @@ Methods
+   .. note::
+     - The ESP32 port does not support the option hard=True.
+
+     - At the RP2040 port UART.IRQ_TXIDLE is only triggered when the message
```
I suggest wording like "The rp2 port's UART.IRQ_TXIDLE is ...".

robert-hh replied:
Done

dpgeorge commented on docs/library/machine.UART.rst:
```diff
@@ -203,11 +178,85 @@ Methods
+       is longer than 5 characters and the trigger happens when still 5 characters
+       are to be sent.
+
+     - At the RP2040 port UART.IRQ_BREAK needs receiving valid characters for triggering
```
Suggest "The rp2 port's UART.IRQ_BREAK needs ...".

robert-hh replied:
Done

dpgeorge commented on docs/library/machine.UART.rst:
```diff
@@ -203,11 +178,85 @@ Methods
+     - At the RP2040 port UART.IRQ_BREAK needs receiving valid characters for triggering
+       again.
+
+     - At the SAMD port UART.IRQ_TXIDLE is triggered while the last character is sent.
```
"The samd port's ..."

robert-hh replied:
Changed

dpgeorge commented on docs/library/machine.UART.rst:
```diff
@@ -203,11 +178,85 @@ Methods
+
+     - At the SAMD port UART.IRQ_TXIDLE is triggered while the last character is sent.
+
+     - At the STM32F4xx using the trigger UART.IRQ_RXIDLE the handler will be called once
```
"On STM32F4xx MCU's, using the trigger ..."

robert-hh replied:
changed

dpgeorge commented on docs/library/machine.UART.rst:
```diff
@@ -203,11 +178,85 @@ Methods
+       idle.
+
+
+   Availability: cc3200, esp32, mimxrt, nrf, renesas-ra, rp2040, samd, stm32.
```
rp2040 -> rp2

robert-hh replied:
changed

dpgeorge commented on docs/library/machine.UART.rst:
```diff
@@ -203,11 +178,85 @@ Methods
+    IRQ trigger sources.
 
-    Availability: WiPy.
+    Availability: renesas-ra, stm32, esp32, rp2040, mimxrt, samd, cc3200.
```
rp2040 -> rp2

robert-hh replied:
changed

robert-hh commented:
I made all changes, but will a test run with all tests and board again.

Gadgetoid commented:
> * Patched by `setting hard=True` in these cases.

AFAIK the PPP module does not set `hard=True`:

https://github.com/micropython/micropython/blob/654711c699e9f27d37442f10ee01827662c4a2bf/extmod/network_ppp_lwip.c#L230-L235

Should it? Or is the PPP `poll()` method likely to do something verboten by hardware interrupts?

robert-hh commented:
Most likely that is not suitable. In hard irq handlers allocating memory is not possible.

Gadgetoid commented:
There's no reference held to the receieve buffer:

https://github.com/micropython/micropython/blob/0d313fb721cdd2516cf3c65333f5d89c6ce632d2/ports/mimxrt/machine_uart.c#L339-L340

So if you add a `gc.collect()` to the txidle test, it will crash.

Minimal repro for RP2:

```python
# Test machine.UART.IRQ_TXIDLE firing after a transmission.
# Does not require any external connections.

from machine import UART
import time, sys
import gc

# Configure pins based on the target.
hard_irq = False

uart_id = 0
tx_pin = "GPIO0"
rx_pin = "GPIO1"
min_window = 1
hard_irq = True


def irq(u):
    print("IRQ_TXIDLE:", u.irq().flags() == u.IRQ_TXIDLE)


text = "Hello World" * 30

# Test that the IRQ is called after the write has completed.
for bits_per_s in (2400, 9600, 115200):
    uart = UART(uart_id, bits_per_s, tx=tx_pin, rx=rx_pin, rxbuf=512)
    uart.irq(irq, uart.IRQ_TXIDLE, hard=True)
    
    # Comment me in/out for crashy funtimes!
    gc.collect()

    bits_per_char = 10  # 1(startbit) + 8(bits) + 1(stopbit) + 0(parity)
    start_time_us = (len(text) - 10) * bits_per_char * 1_000_000 // bits_per_s
    window_ms = max(min_window, 20 * bits_per_char * 1_000 // bits_per_s + 1)

    print("write", bits_per_s)
    uart.write(text)
    time.sleep_us(start_time_us)
    print("ready")
    time.sleep_ms(window_ms)
    print("done")

```

robert-hh commented:
About the NRF port: testing with a Seeed XIAO NRF52840 instead of the Arduino Nano 33 BLE, test batches started by run-tests.py work,  but the test for IRQ_TXIDLE fails if the rx/tx loopback is present.

robert-hh commented:
> So if you add a gc.collect() to the txidle test, it will crash.

The crash is not related to `hard=True`. Let's look for a free floating pointer.

Gadgetoid commented:
> > So if you add a gc.collect() to the txidle test, it will crash.
> 
> The crash is not related to `hard=True`. Let's look for a free floating pointer.

You mean this one?  😆 

https://github.com/micropython/micropython/blob/0d313fb721cdd2516cf3c65333f5d89c6ce632d2/ports/mimxrt/machine_uart.c#L339-L340

Sorry crossed wires, this is *another* thing I ran into while monkeying with PPP.

robert-hh commented:
I found the culprit. It is the mp_irq_obj pointer. Since the RP2 has static UART objects, these were not linked to root pointers and a gc.collect() would free them.

Gadgetoid commented:
Awesome! And here I am still staring glassy eyed at ring buffer code chasing one of those read herons.... who would read a heron anyway?!

robert-hh commented:
@Gadgetoid I updated the RP2 file yesterday evening to address the crash with gc.collect(). SO you can try the most recent version of the PR.
@dpgeorge There was an issue with non-rooted mp_irq_obj objects and the fact, that with RP2 and NRF the UART objects are static.  That is fixed now. It did not reduce the behavior differences of the two NRF boards I have.

robert-hh committed:
nrf/modules/machine/uart: Allow changing the UART baud rate w/o reset.

This commit fixes a bug in the existing driver, that the UART baud rate
could not be changed without reset or power cycle.  It adds as well
functionality to UART.deinit().

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
@dpgeorge There is a new commit now which fixes a bug in the NRF UART, which did not allow to re-instantiate the UART with a different baud rate. Changing the baud rate required a reset. I can squash this commit with the other NRF commit, or even make a separate PR, if you prefer.

robert-hh commented:
Tested with a RPI PICO2 in both ARM and RISCV mode. The test cases pass.

dpgeorge commented on ports/rp2/machine_uart.c:
```diff
@@ -422,6 +448,9 @@ static mp_obj_t mp_machine_uart_make_new(const mp_obj_type_t *type, size_t n_arg
     // Get static peripheral object.
     machine_uart_obj_t *self = (machine_uart_obj_t *)&machine_uart_obj[uart_id];
+    self->mp_irq_obj = NULL;
+    MP_STATE_PORT(rp2_uart_irq_obj[self->uart_id]) = self->mp_irq_obj;
```
Does this irq obj need to be cleared?  Usually something like `machine.UART(1)` is used to retrieve an existing object and not reset it.  But here, the IRQ will be reset (actually the hardware IRQ stays active, if it's enabled).

robert-hh replied:
The behavior of `machine.UART(1)` is not consistent over the  ports. Ig a port has static UART objects like RP2 or NRF, it will refer to the existing state. If a port has dynamic objects like STM32, ESP32, MIMXRT, Renesas, SAMD,  will instantiate a new object. It seems to be more consistent to see it as a new object and clear previous states.

dpgeorge replied:
> If a port has static UART objects like RP2 or NRF, it will refer to the existing state.

Yes, exactly, on rp2 it will refer to the existing instance.  And so in this case (eg accessing `machine.UART(1)`) the IRQ settings should not be changed, for consistency.

> If a port has dynamic objects like STM32, ESP32, MIMXRT, Renesas, SAMD, will instantiate a new object.

Actually, on stm32 any previous UART instance is also retrieved unchanged with a `machine.UART(1)`.

> Yes, it needs to be cleared to have a clean state after soft reset

Then that probably needs to be done in a `machine_uart_deinit()` function.

robert-hh replied:
It can indeed be done there. Thanks for the hint. But uart.deinit() is not automatically called on soft reset.

robert-hh replied:
I could call `machine_uart_deinit()` in `mp_machine_uart_make_new` to settle the ground.

dpgeorge replied:
I would think the UART can behave the same way as Pin, in the way IRQs are handled, and the way they are deinitialised on soft reset.

robert-hh replied:
So I added a `machine_uart_deinit_all()`  function and call it in main() upon soft reset,
and cleared the uart.irq() state in uart.deinit().

dpgeorge replied:
Great, this is much better, thanks!

dpgeorge commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -293,6 +335,58 @@ static bool mp_machine_uart_txdone(machine_uart_obj_t *self) {
+    if (self->mp_irq_obj == NULL) {
+        self->mp_irq_trigger = 0;
+        self->mp_irq_obj = mp_irq_new(&uart_irq_methods, MP_OBJ_FROM_PTR(self));
+        MP_STATE_PORT(nrf_uart_irq_obj) = self->mp_irq_obj;
```
In case one day more UARTs are supported by this port, this should be generalised to an array, and do something like:
```c
MP_STATE_PORT(nrf_uart_irq_obj)[self->uart_id] = self->mp_irq_obj;
```

robert-hh replied:
There is no uart_id element in the `machine_uart_obj_t` data type. I can add it, or use a constant index of 0.

robert-hh replied:
Added uart_id.

dpgeorge commented on tests/extmod/machine_uart_irq_txidle.py:
```diff
@@ -0,0 +1,82 @@
+    if hard_irq is True:
+        uart.irq(irq, uart.IRQ_TXIDLE, hard=True)
+    else:
+        uart.irq(irq, uart.IRQ_TXIDLE)
```
It should be possible to do `uart.irq(irq, uart.IRQ_TXIDLE, hard=hard_irq)`, unconditionally.

robert-hh replied:
The CC200 port for instance does not support the `hard=xxx` argument. But CC3200 does not provide the IRQ_TXIDLE event, so the test will be skipped.

robert-hh replied:
Nevertheless changed.

dpgeorge commented on tests/extmod/machine_uart_irq_txidle.py:
```diff
@@ -0,0 +1,82 @@
+import time, sys
+
+# Configure pins based on the target.
+hard_irq = False
```
Maybe add a comment: "use hard IRQ on ports that need to respond immediately to catch the TXIDLE flag"

robert-hh replied:
OK.

dpgeorge replied:
Ideally we could remove these kinds of small inconsistencies between ports and make the IRQ behave as similar as possible across the ports.

Just to be clear, the hard IRQ is needed only on rp2 when TX/RX have a connection between them (which isn't required for this test but is for the other ones)?

And for samd?

Maybe would be good to make the comment here more detailed, with the above details, so someone who wants to eventually fix things can know exactly what to look for, what tests to run to check if hard IRQ is really needed.

robert-hh replied:
> Just to be clear, the hard IRQ is needed only on rp2 when TX/RX have a connection between them (which isn't required for this test but is for the other ones)?

Yes. The NRF port has similar issues, but does not support hard irq. The SAMD port is fine.

robert-hh replied:
I assume that the IRQ_TXIDLE is triggered while still data is sent. For RP2 this is doumented as 4-5 bytes, but for SAMD21 this is true for one byte as well. Then, when the IRQ callback is called, the IRQ_TXIDLE flag of self->irq_flags is overwritten by the byte TX interrupt.

dpgeorge replied:
> we could change the NRF port to accept & ignore the hard irq argument instead raising an error.

All ports should (eventually) be using the shared code in `shared/runtime/mpirq.c`, and then they would all support the `hard` argument (I guess).  But getting all this consistent now may be too big of a job and we'd never finish :)

> Then, when the IRQ callback is called, the IRQ_TXIDLE flag of self->irq_flags is overwritten by the byte TX interrupt.

So maybe there's something to fix here: if the user doesn't want to respond to TX IRQ's (there's no such thing as far as the user is concerned) then `self->irq_flags` should not be updated when this IRQ occurs.

robert-hh replied:
> All ports should (eventually) be using the shared code in shared/runtime/mpirq.c, and then they would all support the hard argument (I guess). 

The NRF port uses the code from  shared/runtime/mpirq.c but raises an error if the `hard` argument is used.

> if the user doesn't want to respond to TX IRQ's (there's no such thing as far as the user is concerned) then self->irq_flags should not be updated when this IRQ occurs.

With the RX/TX loopback it's the RX event which is triggered. And the uart.irq() callback and the uart IRQ handler are not synchronized. Ideally the respective flag for an event would only be cleared if the callback has finished. But that is not possible for a scheduled callback.

robert-hh replied:
I reworked the handling of the the flags reported by irq.flags(). It will only be changed when a event is present which is expected by the trigger. With that change, hard=True is not needed for the TXIDLE test, and the RX/TX loopback does not cause any problem.

dpgeorge replied:
Excellent!

robert-hh commented on ports/rp2/machine_uart.c:
```diff
@@ -460,6 +489,77 @@ static void mp_machine_uart_sendbreak(machine_uart_obj_t *self) {
+    if (self->mp_irq_obj == NULL) {
+        self->mp_irq_trigger = 0;
+        self->mp_irq_obj = mp_irq_new(&uart_irq_methods, MP_OBJ_FROM_PTR(self));
+        MP_STATE_PORT(rp2_uart_irq_obj[self->uart_id]) = self->mp_irq_obj;
```
Done. But there are more and older places where the above bracket placement is used. Should these be changed as well? Looking into the generated code, I did not see a difference.

dpgeorge replied:
> But there are more and older places where the above bracket placement is used. Should these be changed as well?

No, leave them, we can clean them up separately (separate PR).  That's low priority though.



> Looking into the generated code, I did not see a difference.

The generated code is indeed equivalent.  But it's more conceptual.

dpgeorge commented on ports/esp32/machine_timer.c:
```diff
@@ -130,7 +112,33 @@ static mp_obj_t machine_timer_make_new(const mp_obj_type_t *type, size_t n_args,
 }
 
-static void machine_timer_disable(machine_timer_obj_t *self) {
+machine_timer_obj_t *machine_timer_create(mp_uint_t timer) {
```
I think this new create function should be used by the above `machine_timer_make_new`, because otherwise there's a lot of duplicated code.

robert-hh replied:
Done.

robert-hh replied:
For ESP32, SAMD, Renesas-RA and NRf there are still two commits. I can join them, leaving just one commit per port.

dpgeorge replied:
Thanks for cleaning up the esp32 timer functions.  The diff is smaller now, which is great!

dpgeorge commented on ports/mimxrt/hal/fsl_lpuart.c:
```diff
@@ -0,0 +1,2013 @@
+/*
```
Which MCU source file is this taken from?  Doesn't it need to be different for each MCU?

Is there a better way to fix the code here rather than copying the file?

robert-hh replied:
As far as I recall, that was from the 1062 lib file. But this file is identical for all MIMXRT1xxx MCUs. Newer Versions of the lib use a common UART driver file for all MIMXRT MCU, and this hal/fsl_lpuart.c file is besisdes the changes identical to the one I used a baseline.
I tested it with different MCUs, but can re-test it again.
The change was needed to make IRQ_RXIDLE work as intended by the MCU developers.

dpgeorge replied:
If you tested it on different MCUs then I'm happy with that.

robert-hh replied:
I repeated the test with MIMXRT1010, MIMXRT1015, MIMXRT1021, MIMXRT1052, MIMXRT1062 and MIMXRT1172. All work fine. The test script is set up for Teensy 4.x, but that restriction can be dropped, since the MIMXRT port does not need to specify UART pins.

dpgeorge replied:
Very good.

andrewleech replied:
Hi @robert-hh I'm looking at this file from the perspective of pulling in a new version of it for SDK update. With the auto-formatting done to micropython code it's hard to be sure of the changes you've made.
Was it just the changes in `LPUART_TransferHandleIRQ()` where the large block in the middle `/* If IDLE flag is set and the IDLE interrupt is enabled. */` was removed and replaced with the smaller version at the end?

Before:
``` c
    /* If IDLE flag is set and the IDLE interrupt is enabled. */
    if ((0U != ((uint32_t)kLPUART_IdleLineFlag & status)) &&
        (0U != ((uint32_t)kLPUART_IdleLineInterruptEnable & enabledInterrupts)))
    {
#if defined(FSL_FEATURE_LPUART_HAS_FIFO) && FSL_FEATURE_LPUART_HAS_FIFO
        count = ((uint8_t)((base->WATER & LPUART_WATER_RXCOUNT_MASK) >> LPUART_WATER_RXCOUNT_SHIFT));

        while ((0U != handle->rxDataSize) && (0U != count))
        {
            tempCount = (uint8_t)MIN(handle->rxDataSize, count);

            /* Using non block API to read the data from the registers. */
            LPUART_ReadNonBlocking(base, handle->rxData, tempCount);
            handle->rxData = &handle->rxData[tempCount];
            handle->rxDataSize -= tempCount;
            count -= tempCount;

            /* If rxDataSize is 0, invoke rx idle callback.*/
            if (0U == (handle->rxDataSize))
            {
                handle->rxState = (uint8_t)kLPUART_RxIdle;

                if (NULL != handle->callback)
                {
                    handle->callback(base, handle, kStatus_LPUART_RxIdle, handle->userData);
                }
            }
        }
#endif
        /* Clear IDLE flag.*/
        base->STAT = ((base->STAT & 0x3FE00000U) | LPUART_STAT_IDLE_MASK);

        /* If rxDataSize is 0, disable rx ready, overrun and idle line interrupt.*/
        if (0U == handle->rxDataSize)
        {
            /* Disable and re-enable the global interrupt to protect the interrupt enable register during
             * read-modify-wrte. */
            irqMask = DisableGlobalIRQ();
            base->CTRL &= ~(uint32_t)(LPUART_CTRL_RIE_MASK | LPUART_CTRL_ILIE_MASK | LPUART_CTRL_ORIE_MASK);
            EnableGlobalIRQ(irqMask);
        }
        /* Invoke callback if callback is not NULL and rxDataSize is not 0. */
        else if (NULL != handle->callback)
        {
            handle->callback(base, handle, kStatus_LPUART_IdleLineDetected, handle->userData);
        }
        else
        {
            /* Avoid MISRA 15.7 */
        }
    }
```
after 
``` c
    /* If IDLE flag is set and the IDLE interrupt is enabled. */
    if ((0U != ((uint32_t)kLPUART_IdleLineFlag & status)) &&
        (0U != ((uint32_t)kLPUART_IdleLineInterruptEnable & enabledInterrupts))) {
        /* Clear IDLE flag.*/
        base->STAT |= LPUART_STAT_IDLE_MASK;
        if (NULL != handle->callback) {
            handle->callback(base, handle, kStatus_LPUART_IdleLineDetected, handle->userData);
        } else {
            /* Avoid MISRA 15.7 */
        }
    }
```

robert-hh replied:
Comparinf is easier if you run the initial fsl_lpuart.c through uncrustify as well before comparing. So yes, the changes focus on handling the line_idle signal. I removed most of the initial code dealing with it and just kept the part which handles the callback mechanism. So it looks like the code you show is identical to the blocks below.
So this block is removed, which is an `if() .....`:
```
    /* If IDLE flag is set and the IDLE interrupt is enabled. */
    if ((0U != ((uint32_t)kLPUART_IdleLineFlag & status)) &&
        (0U != ((uint32_t)kLPUART_IdleLineInterruptEnable & enabledInterrupts))) {
        #if defined(FSL_FEATURE_LPUART_HAS_FIFO) && FSL_FEATURE_LPUART_HAS_FIFO
        count = ((uint8_t)((base->WATER & LPUART_WATER_RXCOUNT_MASK) >> LPUART_WATER_RXCOUNT_SHIFT));

        while ((0U != handle->rxDataSize) && (0U != count)) {
            tempCount = (uint8_t)MIN(handle->rxDataSize, count);

            /* Using non block API to read the data from the registers. */
            LPUART_ReadNonBlocking(base, handle->rxData, tempCount);
            handle->rxData = &handle->rxData[tempCount];
            handle->rxDataSize -= tempCount;
            count -= tempCount;

            /* If rxDataSize is 0, invoke rx idle callback.*/
            if (0U == (handle->rxDataSize)) {
                handle->rxState = (uint8_t)kLPUART_RxIdle;

                if (NULL != handle->callback) {
                    handle->callback(base, handle, kStatus_LPUART_RxIdle, handle->userData);
                }
            }
        }
        #endif
        /* Clear IDLE flag.*/
        base->STAT = ((base->STAT & 0x3FE00000U) | LPUART_STAT_IDLE_MASK);

        /* If rxDataSize is 0, disable rx ready, overrun and idle line interrupt.*/
        if (0U == handle->rxDataSize) {
            /* Disable and re-enable the global interrupt to protect the interrupt enable register during
             * read-modify-wrte. */
            irqMask = DisableGlobalIRQ();
            base->CTRL &= ~(uint32_t)(LPUART_CTRL_RIE_MASK | LPUART_CTRL_ILIE_MASK | LPUART_CTRL_ORIE_MASK);
            EnableGlobalIRQ(irqMask);
        }
        /* Invoke callback if callback is not NULL and rxDataSize is not 0. */
        else if (NULL != handle->callback) {
            handle->callback(base, handle, kStatus_LPUART_IdleLineDetected, handle->userData);
        } else {
            /* Avoid MISRA 15.7 */
        }
    }
```
And this on is added, which was part of the removed block:
```
    /* If IDLE flag is set and the IDLE interrupt is enabled. */
    if ((0U != ((uint32_t)kLPUART_IdleLineFlag & status)) &&
        (0U != ((uint32_t)kLPUART_IdleLineInterruptEnable & enabledInterrupts))) {
        /* Clear IDLE flag.*/
        base->STAT |= LPUART_STAT_IDLE_MASK;
        if (NULL != handle->callback) {
            handle->callback(base, handle, kStatus_LPUART_IdleLineDetected, handle->userData);
        } else {
            /* Avoid MISRA 15.7 */
        }
    }
```

dpgeorge commented:
@robert-hh this is looking really good now, thank you for all the effort on it, especially with the testing.

Just two remaining comments, see above.

robert-hh commented:
The change to esp32/machine_timer.c is made and the MIMXRT MCU's re-tested. I removed at the test scripts the restrictions for Teensy 4.x. So any MIMXRT board can be used.

dpgeorge commented:
> I removed at the test scripts the restrictions for Teensy 4.x. So any MIMXRT board can be used.

OK, great.  I also removed the test configuration for the samd `SEEED_XIAO_SAMD21` because that doesn't have UART IRQs enabled by default (I was enabling them manually to test on this boad).  I have now ordered an `ADAFRUIT_ITSYBITSY_M0_EXPRESS` to use for testing instead.

I also did some very minor fixes to this PR:
- remove some stray double blank lines
- add a few blank lines around `#if`/`#endif`'s
- make `const mp_irq_methods_t uart_irq_methods` static

And I kept all 16 commits.  They are well separated commits making independent changes.

Thanks @robert-hh for all your work on this!  And @Gadgetoid for testing, and @projectgus for additional reviews.

robert-hh commented:
Thank you for the improvements and for merging.

iabdalkader commented:
Note this doesn't build anymore with more recent SDKs because the isr variable is now defined:
```
extern lpuart_isr_t s_lpuartIsr[];
```

robert-hh commented:
> Note this doesn't build anymore with more recent SDKs

Which port?

iabdalkader commented:
> Which port?

mimxrt, with any recent SDK release.

robert-hh commented:
We are still at sdk 2.10, and we use a modified fsl_lpuart.c to get a proper IRQ_RXIDLE behavior. Updating to a newer SDK is still due. When that happens I expect more changes to be done. No too many, as I noticed when trying that last year.

iabdalkader commented:
> We are still at sdk 2.10, and we use a modified fsl_lpuart.c to get a proper IRQ_RXIDLE behavior.

I understand. I'm just saying if you try to use a more recent SDK this doesn't build anymore. However, it seems everything else works. I'm not sure which SDK I'm using but it's more recent than the one in MicroPython.
